From 3131c40c636f4c39b729cae4cbe9f909efa86ddc Mon Sep 17 00:00:00 2001
From: Sudheesh Mavila <sudheesh.mavila@amd.com>
Date: Tue, 12 Mar 2019 22:33:11 +0530
Subject: [PATCH 1/4] amd-xgbe Sometimes driver report incorrect link status
 and results in timeout error in driver EMBSWDEV-6825

Upstream-Status: Pending
Patch from vendor in attachment of https://jira.wrs.com/browse/LINUXPG-431

Signed-off-by: Sudheesh Mavila <sudheesh.mavila@amd.com>
Signed-off-by: Yongxin Liu <yongxin.liu@windriver.com>
---
 drivers/net/ethernet/amd/xgbe/xgbe-phy-v2.c | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/amd/xgbe/xgbe-phy-v2.c b/drivers/net/ethernet/amd/xgbe/xgbe-phy-v2.c
index 213769054391..0e2215273f09 100644
--- a/drivers/net/ethernet/amd/xgbe/xgbe-phy-v2.c
+++ b/drivers/net/ethernet/amd/xgbe/xgbe-phy-v2.c
@@ -2628,8 +2628,19 @@ static int xgbe_phy_link_status(struct xgbe_prv_data *pdata, int *an_restart)
 	 */
 	reg = XMDIO_READ(pdata, MDIO_MMD_PCS, MDIO_STAT1);
 	reg = XMDIO_READ(pdata, MDIO_MMD_PCS, MDIO_STAT1);
-	if (reg & MDIO_STAT1_LSTATUS)
-		return 1;
+
+	if(phy_data->sfp_speed == XGBE_SFP_SPEED_10000) {
+		if ((reg & MDIO_STAT1_LSTATUS) && !(reg & MDIO_STAT1_FAULT)) {
+			return 1;
+		} else {
+			*an_restart = 1;
+			pdata->phy_if.phy_reset(pdata);
+			return 0;
+		}
+	} else {
+		if (reg & MDIO_STAT1_LSTATUS)
+			return 1;
+	}
 
 	if (pdata->phy.autoneg == AUTONEG_ENABLE &&
 	    phy_data->port_mode == XGBE_PORT_MODE_BACKPLANE) {
-- 
2.34.1

